build/upload-coverage.sh

```diff
@@ -1,80 +1,52 @@
 #!/usr/bin/env bash
 
-set -euo pipefail
+set -euxo pipefail
 
-if [ -z "$COVERALLS_TOKEN" ]; then
+if [ -z "${COVERALLS_TOKEN-}" ]; then
   echo "FAIL: Missing or empty COVERALLS_TOKEN."
   exit 1
 fi
-if [ -z "$CODECOV_TOKEN" ]; then
+if [ -z "${CODECOV_TOKEN-}" ]; then
   echo "FAIL: Missing or empty CODECOV_TOKEN."
   exit 1
 fi
 
-if [ -n "${TMPDIR-}" ]; then
-  outdir="${TMPDIR}"
-else
-  outdir="/tmp"
-fi
-
-prefix="github.com/cockroachdb/cockroach/pkg"
-# This regex removes files from the uploaded coverage.
-ignore_files="$prefix/(acceptance|cmd|ui/embedded|storage/simulation|sql/pgbench|.*\.(pb|pb\.gw)\.go)"
+prefix=github.com/cockroachdb/cockroach/pkg
 
-coverage_dir="${outdir}/coverage"
-coverage_profile="${coverage_dir}/coverage.out"
-coverage_mode=count
+coverage_dir=coverage
+coverage_profile=$coverage_dir/coverage.out
 
-# iterative_coverpkg fetches all test deps and main deps, filters them, and
-# converts them into a comma separated list stored in $coverpkg.
-iterative_coverpkg() {
-  local imports="$1"
-  local old_line_count="-1"
-  local line_count=""
-  while [ "$old_line_count" != "$line_count" ]; do
-    old_line_count=$line_count
-    imports+=$'\n'$(go list  -f '{{join .Imports "\n"}}
-{{join .TestImports "\n"}}
-{{join .XTestImports "\n"}}' $imports | grep $prefix)
-    imports=$(echo "$imports" | sort | uniq)
-    line_count=$(echo $imports | wc -w)
-  done
-  coverpkg=$(echo $imports | sed 's/ /,/g')
-}
+trap "rm -rf $coverage_dir" EXIT
 
-rm -rf "$coverage_dir"
-mkdir -p "$coverage_dir"
-
-# Run "make coverage" on each package except for those in cmd.
-for pkg in $(go list ./pkg/... | grep -vF 'pkg/cmd'); do
-  echo "Processing $pkg..."
-  # Verify package has test files.
-  if [ -z "$(go list -f '{{join .TestGoFiles ""}}{{join .XTestGoFiles ""}}' $pkg)" ]; then
-    echo "$pkg: Skipping due to no test files."
-    continue
-  fi
+rm -rf $coverage_dir
+mkdir -p $coverage_dir
 
+# Run coverage on each package, because go test "cannot use test profile flag
+# with multiple packages".
+for pkg in $(go list $prefix/...); do
   # Only generate coverage for cockroach dependencies.
-  iterative_coverpkg $pkg
-
-  # Find all cockroach packages that are imported by this package or in it's
-  # tests.
-  f="${coverage_dir}/$(echo $pkg | tr / -).cover"
-  touch $f
-  time make coverage \
-    PKG="$pkg" \
-    TESTFLAGS="-v -coverprofile=$f -covermode=$coverage_mode -coverpkg=$coverpkg" | \
-    tee "${outdir}/coverage.log"
+  #
+  # Note that grep's exit status is ignored here to allow for packages with no
+  # dependencies.
+  coverpkg=$(go list -f '{{join .Imports "\n"}}{{"\n"}}{{join .TestImports "\n"}}{{"\n"}}{{join .XTestImports "\n"}}' $pkg | \
+    sort | uniq | grep -v '^C$' | \
```

Comment:
Is there any DRYing we can do between this and the other one? Seems like maybe not, but you're in the best position to answer.